tests/test_packages/test_connections/test_p2p_libp2p/test_communication.py

```diff
@@ -668,3 +669,110 @@ def teardown(self):
             shutil.rmtree(self.t)
         except (OSError, IOError):
             pass
+
+
+@libp2p_log_on_failure_all
+class TestP2PLibp2pReconnectionSendEnvelope:
+    """Test that connection will send envelope with error, and that reconnection fixes it."""
+
+    @classmethod
+    @libp2p_log_on_failure
+    def setup_class(cls):
+        """Set the test up"""
+        cls.cwd = os.getcwd()
+        cls.t = tempfile.mkdtemp()
+        os.chdir(cls.t)
+
+        cls.log_files = []
+        cls.multiplexers = []
+
+        aea_ledger_fetchai = make_crypto(FetchAICrypto.identifier)
+        aea_ledger_ethereum = make_crypto(EthereumCrypto.identifier)
+
+        try:
+            temp_dir_1 = os.path.join(cls.t, "temp_dir_1")
+            os.mkdir(temp_dir_1)
+            cls.connection1 = _make_libp2p_connection(
+                data_dir=temp_dir_1, agent_key=aea_ledger_fetchai, port=DEFAULT_PORT + 1
+            )
+            cls.multiplexer1 = Multiplexer(
+                [cls.connection1], protocols=[MockDefaultMessageProtocol]
+            )
+            cls.log_files.append(cls.connection1.node.log_file)
+            cls.multiplexer1.connect()
+            cls.multiplexers.append(cls.multiplexer1)
+
+            genesis_peer = cls.connection1.node.multiaddrs[0]
+
+            temp_dir_2 = os.path.join(cls.t, "temp_dir_2")
+            os.mkdir(temp_dir_2)
+            cls.connection2 = _make_libp2p_connection(
+                data_dir=temp_dir_2,
+                port=DEFAULT_PORT + 2,
+                entry_peers=[genesis_peer],
+                agent_key=aea_ledger_ethereum,
+            )
+            cls.multiplexer2 = Multiplexer(
+                [cls.connection2], protocols=[MockDefaultMessageProtocol]
+            )
+            cls.log_files.append(cls.connection2.node.log_file)
+            cls.multiplexer2.connect()
+            cls.multiplexers.append(cls.multiplexer2)
+        except Exception as e:
+            cls.teardown_class()
+            raise e
+
+    def test_connection_is_established(self):
```

Comment:
Remember, tests in a class are executed in random order

Reply:
yes